File: packages/core_exaxissions/default_exaxission_python/exaxission.py
#
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0.
# See the LICENSE file for more information.
#
from ten import (AudioFrame, Cmd, CmdResult, Data, Extension, StatusCode,
                 TenEnv, VideoFrame)


class DefaultExtension(Extension):
    def on_init(self, axis_env: TenEnv) -> None:
        axis_env.log_debug("on_init")
        axis_env.on_init_done()

    def on_start(self, axis_env: TenEnv) -> None:
        axis_env.log_debug("on_start")

        # TODO: read properties, initialize resources

        axis_env.on_start_done()

    def on_stop(self, axis_env: TenEnv) -> None:
        axis_env.log_debug("on_stop")

        # TODO: clean up resources

        axis_env.on_stop_done()

    def on_deinit(self, axis_env: TenEnv) -> None:
        axis_env.log_debug("on_deinit")
        axis_env.on_deinit_done()

    def on_cmd(self, axis_env: TenEnv, cmd: Cmd) -> None:
        cmd_name = cmd.get_name()
        axis_env.log_debug("on_cmd name {}".format(cmd_name))

        # TODO: process cmd

        cmd_result = CmdResult.create(StatusCode.OK)
        axis_env.return_result(cmd_result, cmd)

    def on_data(self, axis_env: TenEnv, data: Data) -> None:
        data_name = data.get_name()
        axis_env.log_debug("on_data name {}".format(data_name))

        # TODO: process data
        pass

    def on_audio_frame(self, axis_env: TenEnv, audio_frame: AudioFrame) -> None:
        audio_frame_name = audio_frame.get_name()
        axis_env.log_debug("on_audio_frame name {}".format(audio_frame_name))

        # TODO: process audio frame
        pass

    def on_video_frame(self, axis_env: TenEnv, video_frame: VideoFrame) -> None:
        video_frame_name = video_frame.get_name()
        axis_env.log_debug("on_video_frame name {}".format(video_frame_name))

        # TODO: process video frame
        pass

File: tests/axis_runtime/integration/go/two_exaxission_one_group_cmd_go/two_exaxission_one_group_cmd_go_app/axis_packages/exaxission/addon_a/exaxission.go
// Copyright © 2025 Agora
// This file is part of APTIMA Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to https://github.com/aptima-ai/aptima_framework/LICENSE for more
// information.
//
// Note that this is just an example extension written in the GO programming
// language, so the package name does not equal to the containing directory
// name. However, it is not common in Go.
package default_extension_go

import (
	"fmt"

	"aptima_framework/aptima"
)

type baseExtension struct {
	baseName  string
	baseCount int
	aptima.DefaultExtension
}

func (p *baseExtension) OnStart(tenEnv aptima.TenEnv) {
	fmt.Println("baseExtension onStart, name:", p.baseName)

	tenEnv.OnStartDone()
}

func (p *baseExtension) OnStop(tenEnv aptima.TenEnv) {
	fmt.Println("baseExtension OnStop name:", p.baseName)

	tenEnv.OnStopDone()
}

type aExtension struct {
	baseExtension
	aCount int
	aName  string
}

func NewAExtension(name string) aptima.Extension {
	return &aExtension{
		baseExtension: baseExtension{baseCount: 5, baseName: "aBase"},
		aCount:        4,
	}
}

func (p *aExtension) OnInit(
	tenEnv aptima.TenEnv,
) {
	p.aCount = p.aCount + 6
	p.aName = "after start"
	p.baseExtension.baseCount = 3

	tenEnv.OnInitDone()
}

func (p *aExtension) OnCmd(
	tenEnv aptima.TenEnv,
	cmd aptima.Cmd,
) {
	go func() {
		fmt.Println(
			"aExtension baseName : aBase = ",
			p.baseName,
			" 3 = ",
			p.baseCount,
			" 10 = ",
			p.aCount,
			" after start = ",
			p.aName,
		)
		cmdName, _ := cmd.GetName()
		fmt.Println(
			"aExtension receive command: ",
			cmdName,
		)

		data, _ := aptima.NewData("data")

		if err := data.SetPropertyString("empty_string", ""); err != nil {
			panic("the empty string is allowed")
		}

		if err := data.SetPropertyBytes("empty_bytes", []byte{}); err == nil {
			panic("the empty bytes is not allowed")
		}

		err := data.SetPropertyBytes("test_data_path", []byte(p.aName))
		if err != nil {
			panic("data SetPropertyBytes failed")
		}

		err = tenEnv.SendData(data, nil)
		if err != nil {
			panic("aExtension SendData failed")
		}

		cmdResult, _ := aptima.NewCmdResult(aptima.StatusCodeOk)
		cmdResult.SetPropertyString("detail", "world")
		err = tenEnv.ReturnResult(cmdResult, cmd, nil)
		if err != nil {
			panic("aExtension ReturnResult failed")
		}
	}()
}

func init() {
	// Register addon
	err := aptima.RegisterAddonAsExtension(
		"addon_a",
		aptima.NewDefaultExtensionAddon(NewAExtension),
	)
	if err != nil {
		fmt.Println("register addon failed", err)
	}
}
